// @wdesgardin/fivem-js/lib/weaponComponent/WeaponComponentHudStats.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeaponComponentHudStats = void 0;
const utils_1 = require("../utils");
const WeaponComponentHash_1 = require("./WeaponComponentHash");
/**
 * Mapping of WeaponComponentHash -> WeaponComponentHudStats
 *
 */
exports.WeaponComponentHudStats = new Map();
function initializeOnce() {
    let isInitialized = false;
    return function () {
        if (isInitialized) {
            return;
        }
        // magic number based on struct WeaponComponentHudStat
        const intLength = 4;
        for (const hash of (0, utils_1.enumValues)(WeaponComponentHash_1.WeaponComponentHash)) {
            const buffer = new Uint8Array(0x28);
            // https://docs.fivem.net/natives/?_0xB3CAF387AE12E9F8
            Citizen.invokeNative('0xB3CAF387AE12E9F8', hash, buffer, Citizen.returnResultAnyway());
            // noinspection PointlessArithmeticExpressionJS
            const weaponComponentHudStat = {
                hudDamage: (0, utils_1.getUInt32FromUint8Array)(buffer, 0 * intLength, 1 * intLength),
                hudSpeed: (0, utils_1.getUInt32FromUint8Array)(buffer, 2 * intLength, 3 * intLength),
                hudCapacity: (0, utils_1.getUInt32FromUint8Array)(buffer, 4 * intLength, 5 * intLength),
                hudAccuracy: (0, utils_1.getUInt32FromUint8Array)(buffer, 6 * intLength, 7 * intLength),
                hudRange: (0, utils_1.getUInt32FromUint8Array)(buffer, 8 * intLength, 9 * intLength),
            };
            exports.WeaponComponentHudStats.set(hash, weaponComponentHudStat);
        }
        isInitialized = true;
    };
}
if ((0, utils_1.IsClient)())
    initializeOnce()();
