// @wdesgardin/fivem-js/lib/weapon/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WeaponAsset = exports.Weapon = exports.WeaponLiveryColor = exports.WeaponLivery = exports.WeaponGroup = exports.WeaponTint = exports.WeaponHudStats = exports.DlcWeaponData = void 0;
var DlcWeaponData_1 = require("./DlcWeaponData");
Object.defineProperty(exports, "DlcWeaponData", { enumerable: true, get: function () { return DlcWeaponData_1.DlcWeaponData; } });
var WeaponHudStats_1 = require("./WeaponHudStats");
Object.defineProperty(exports, "WeaponHudStats", { enumerable: true, get: function () { return WeaponHudStats_1.WeaponHudStats; } });
var WeaponTint_1 = require("./WeaponTint");
Object.defineProperty(exports, "WeaponTint", { enumerable: true, get: function () { return WeaponTint_1.WeaponTint; } });
var WeaponGroup_1 = require("./WeaponGroup");
Object.defineProperty(exports, "WeaponGroup", { enumerable: true, get: function () { return WeaponGroup_1.WeaponGroup; } });
var WeaponLivery_1 = require("./WeaponLivery");
Object.defineProperty(exports, "WeaponLivery", { enumerable: true, get: function () { return WeaponLivery_1.WeaponLivery; } });
var WeaponLiveryColor_1 = require("./WeaponLiveryColor");
Object.defineProperty(exports, "WeaponLiveryColor", { enumerable: true, get: function () { return WeaponLiveryColor_1.WeaponLiveryColor; } });
var Weapon_1 = require("./Weapon");
Object.defineProperty(exports, "Weapon", { enumerable: true, get: function () { return Weapon_1.Weapon; } });
var WeaponAsset_1 = require("./WeaponAsset");
Object.defineProperty(exports, "WeaponAsset", { enumerable: true, get: function () { return WeaponAsset_1.WeaponAsset; } });
