// @wdesgardin/fivem-js/lib/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IsClient = exports.Wait = exports.Vector3 = exports.String = exports.Size = exports.Quaternion = exports.PointF = exports.Point = exports.Maths = exports.LiteEvent = exports.getUInt32FromUint8Array = exports.getStringFromUInt8Array = exports.enumValues = exports.Crypto = exports.Color = void 0;
var Color_1 = require("./Color");
Object.defineProperty(exports, "Color", { enumerable: true, get: function () { return Color_1.Color; } });
var Crypto_1 = require("./Crypto");
Object.defineProperty(exports, "Crypto", { enumerable: true, get: function () { return Crypto_1.Crypto; } });
var enumValues_1 = require("./enumValues");
Object.defineProperty(exports, "enumValues", { enumerable: true, get: function () { return enumValues_1.enumValues; } });
var getStringFromUInt8Array_1 = require("./getStringFromUInt8Array");
Object.defineProperty(exports, "getStringFromUInt8Array", { enumerable: true, get: function () { return getStringFromUInt8Array_1.getStringFromUInt8Array; } });
var getUInt32FromUint8Array_1 = require("./getUInt32FromUint8Array");
Object.defineProperty(exports, "getUInt32FromUint8Array", { enumerable: true, get: function () { return getUInt32FromUint8Array_1.getUInt32FromUint8Array; } });
var LiteEvent_1 = require("./LiteEvent");
Object.defineProperty(exports, "LiteEvent", { enumerable: true, get: function () { return LiteEvent_1.LiteEvent; } });
var Maths_1 = require("./Maths");
Object.defineProperty(exports, "Maths", { enumerable: true, get: function () { return Maths_1.Maths; } });
var Point_1 = require("./Point");
Object.defineProperty(exports, "Point", { enumerable: true, get: function () { return Point_1.Point; } });
var PointF_1 = require("./PointF");
Object.defineProperty(exports, "PointF", { enumerable: true, get: function () { return PointF_1.PointF; } });
var Quaternion_1 = require("./Quaternion");
Object.defineProperty(exports, "Quaternion", { enumerable: true, get: function () { return Quaternion_1.Quaternion; } });
var Size_1 = require("./Size");
Object.defineProperty(exports, "Size", { enumerable: true, get: function () { return Size_1.Size; } });
var String_1 = require("./String");
Object.defineProperty(exports, "String", { enumerable: true, get: function () { return String_1.String; } });
var Vector3_1 = require("./Vector3");
Object.defineProperty(exports, "Vector3", { enumerable: true, get: function () { return Vector3_1.Vector3; } });
const Wait = (milliseconds) => new Promise(resolve => setTimeout(resolve, milliseconds));
exports.Wait = Wait;
const IsClient = () => {
    return typeof IsDuplicityVersion == 'function' && !IsDuplicityVersion();
};
exports.IsClient = IsClient;
