// @wdesgardin/fivem-js/lib/ui/menu/items/panels/AbstractUIMenuPanel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractUIMenuPanel = void 0;
const utils_1 = require("../../../../utils");
const __1 = require("../../../");
const __2 = require("../../");
class AbstractUIMenuPanel {
    constructor() {
        this.id = utils_1.Crypto.uuidv4();
        this.enabled = true;
    }
    get ParentMenu() {
        return this.parentItem ? this.parentItem.parent : undefined;
    }
    get ParentItem() {
        var _a;
        return (_a = this.parentItem) !== null && _a !== void 0 ? _a : undefined;
    }
    set ParentItem(value) {
        this.parentItem = value;
    }
    get Enabled() {
        return this.enabled;
    }
    set Enabled(value) {
        this.enabled = value;
    }
    get Height() {
        return this.background ? this.background.size.height : 0;
    }
    setVerticalPosition(y) {
        if (this.background)
            this.background.pos.Y = y;
    }
    draw() {
        if (this.background) {
            this.background.size.width = 431 + (this.ParentMenu ? this.ParentMenu.WidthOffset : 0);
            this.background.pos.X = this.parentItem ? this.parentItem.offset.X : 0;
            if (this.background instanceof __1.Sprite) {
                this.background.draw(__2.Menu.screenResolution);
            }
            else {
                this.background.draw(undefined, __2.Menu.screenResolution);
            }
        }
    }
}
exports.AbstractUIMenuPanel = AbstractUIMenuPanel;
