// @wdesgardin/fivem-js/lib/enums/GamerTagComponent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GamerTagComponent = void 0;
var GamerTagComponent;
(function (GamerTagComponent) {
    GamerTagComponent[GamerTagComponent["GamerName"] = 0] = "GamerName";
    GamerTagComponent[GamerTagComponent["CrewTag"] = 1] = "CrewTag";
    GamerTagComponent[GamerTagComponent["HealthArmour"] = 2] = "HealthArmour";
    GamerTagComponent[GamerTagComponent["BigText"] = 3] = "BigText";
    GamerTagComponent[GamerTagComponent["AudioIcon"] = 4] = "AudioIcon";
    GamerTagComponent[GamerTagComponent["UsingMenu"] = 5] = "UsingMenu";
    GamerTagComponent[GamerTagComponent["PassiveMode"] = 6] = "PassiveMode";
    GamerTagComponent[GamerTagComponent["WantedStars"] = 7] = "WantedStars";
    GamerTagComponent[GamerTagComponent["Diver"] = 8] = "Diver";
    GamerTagComponent[GamerTagComponent["CoDriver"] = 9] = "CoDriver";
    GamerTagComponent[GamerTagComponent["Tagged"] = 10] = "Tagged";
    GamerTagComponent[GamerTagComponent["GameNameNarby"] = 11] = "GameNameNarby";
    GamerTagComponent[GamerTagComponent["Arrow"] = 12] = "Arrow";
    GamerTagComponent[GamerTagComponent["Packages"] = 13] = "Packages";
    GamerTagComponent[GamerTagComponent["InvIfPedFollowing"] = 14] = "InvIfPedFollowing";
    GamerTagComponent[GamerTagComponent["RankText"] = 15] = "RankText";
    GamerTagComponent[GamerTagComponent["Typing"] = 16] = "Typing";
    GamerTagComponent[GamerTagComponent["BagLarge"] = 17] = "BagLarge";
    GamerTagComponent[GamerTagComponent["Arrow2"] = 18] = "Arrow2";
    GamerTagComponent[GamerTagComponent["GangCeo"] = 19] = "GangCeo";
    GamerTagComponent[GamerTagComponent["GangBiker"] = 20] = "GangBiker";
    GamerTagComponent[GamerTagComponent["BikerArrow"] = 21] = "BikerArrow";
    GamerTagComponent[GamerTagComponent["McRolePresident"] = 22] = "McRolePresident";
    GamerTagComponent[GamerTagComponent["McRoleVicePresident"] = 23] = "McRoleVicePresident";
    GamerTagComponent[GamerTagComponent["McRoleRoadCaptain"] = 24] = "McRoleRoadCaptain";
    GamerTagComponent[GamerTagComponent["McRoleRoadSergent"] = 25] = "McRoleRoadSergent";
    GamerTagComponent[GamerTagComponent["McRoleRoadEnforcer"] = 26] = "McRoleRoadEnforcer";
    GamerTagComponent[GamerTagComponent["McRoleRoadProspect"] = 27] = "McRoleRoadProspect";
    GamerTagComponent[GamerTagComponent["Transmitter"] = 28] = "Transmitter";
    GamerTagComponent[GamerTagComponent["Bomb"] = 29] = "Bomb";
})(GamerTagComponent = exports.GamerTagComponent || (exports.GamerTagComponent = {}));
