// @wdesgardin/fivem-js/lib/GamerTag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GamerTag = void 0;
class GamerTag {
    constructor(handle) {
        this.handle = handle;
    }
    get Handle() {
        return this.handle;
    }
    get IsActive() {
        return !!IsMpGamerTagActive(this.handle);
    }
    get IsFree() {
        return !!IsMpGamerTagFree(this.handle);
    }
    set Name(value) {
        SetMpGamerTagName(this.handle, value);
    }
    set WantedLevel(level) {
        SetMpGamerTagWantedLevel(this.handle, level);
    }
    setHealthBarColour(color) {
        SetMpGamerTagHealthBarColour(this.handle, color);
    }
    setComponentAlpha(component, alpha) {
        SetMpGamerTagAlpha(this.handle, component, alpha);
    }
    setComponentColour(component, hudColorIndex) {
        SetMpGamerTagColour(this.handle, component, hudColorIndex);
    }
    setComponentVisibility(component, visible) {
        SetMpGamerTagVisibility(this.handle, component, visible);
    }
    delete() {
        RemoveMpGamerTag(this.handle);
    }
}
exports.GamerTag = GamerTag;
