// @wdesgardin/fivem-js/lib/models/Player.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Player = void 0;
const _1 = require("./");
class Player {
    constructor(handle) {
        this.pvp = false;
        this.handle = handle;
        this.PvPEnabled = true;
    }
    static fromServerId(serverId) {
        return new Player(GetPlayerFromServerId(serverId));
    }
    get Handle() {
        return this.handle;
    }
    get Character() {
        const handle = GetPlayerPed(this.handle);
        if (typeof this.ped === 'undefined' || handle !== this.ped.Handle) {
            this.ped = new _1.Ped(handle);
        }
        return this.ped;
    }
    get Name() {
        return GetPlayerName(this.handle);
    }
    get PvPEnabled() {
        return this.pvp;
    }
    set PvPEnabled(value) {
        NetworkSetFriendlyFireOption(value);
        SetCanAttackFriendly(this.Character.Handle, value, value);
        this.pvp = value;
    }
    get WantedLevel() {
        return GetPlayerWantedLevel(this.handle);
    }
    set WantedLevel(level) {
        SetPlayerWantedLevel(this.handle, level, false);
        SetPlayerWantedLevelNow(this.handle, false);
    }
    set IgnoredByPolice(value) {
        SetPoliceIgnorePlayer(this.handle, value);
    }
    set IgnoredByEveryone(value) {
        SetEveryoneIgnorePlayer(this.handle, value);
    }
    get ServerId() {
        return GetPlayerServerId(this.handle);
    }
    clearWantedLevel() {
        ClearPlayerWantedLevel(this.handle);
    }
    disableFiringThisFrame() {
        DisablePlayerFiring(this.handle, false);
    }
}
exports.Player = Player;
